kotest-property/src/commonMain/kotlin/io/kotest/property/Gen.kt

```diff
@@ -0,0 +1,112 @@
+package io.kotest.property
+
+import kotlin.random.Random
+
+/**
+ * A Generator, or [Gen] is responsible for generating data to be used in property testing.
+ * Each generator will generate data for a specific type <T>.
+ *
+ * There are two supported types of generators - [Arbitrary] and [Progression] - defined
+ * as sub interfaces of this interface.
+ *
+ * An arbitrary is used when you need random values across a large space.
+ * A progression is useful when you want exhaustive values from a small space.
+ *
+ * Both types of generators can be mixed and matched in property tests. For example,
+ * you could test a function with 1000 random positive integers (arbitrary) and every
+ * even number from 0 to 200 (progression).
+ */
+interface Gen<T> {
+
+   /**
+    * @return the values provided by this [Gen] as a lazy list.
+    * @see [PropertyInput]
+    */
+   fun generate(random: Random): Sequence<PropertyInput<T>>
+}
+
+/**
+ * Models the type of inputs that can be generated from an [Arbitrary].
+ *
+ * There are two implementations of [PropertyInput]. The first is a wrapper for values without
+ * a shrinker, which is [PropertyInput.Value].
+ *
+ * The second is a wrapper for values associated with shrinks, which is [PropertyInput.ValueAndShrinker].
+ */
+sealed class PropertyInput<out T> {
```

Comment:
Shouldn't `PropertyInput` either get it's own file or be with `Arbitrary`?
It's not exactly an `Arbitrary`-only file as stated in the docs, as `Progression` must generate it too (It comes from `Gen`)

Reply:
Done.